File: go/bandwidth/director.go
// SPDX-License-Identifier: MIT

package bandwidth

import (
	"sort"
	"strings"
	"sync"
	"time"
)

// BandwidthDirector is the cross-transport bandwidth synthesis and gossip coordinator.
//
// It maintains a matrix of (peerUhid × transportName) → BandwidthSample estimates
// and provides transport recommendations based on payload size, BDP, and power cost.
//
// Transport selection algorithm:
//  1. Score = AvailableBps / PowerCostRelative (higher is better).
//  2. If payload > BDP: prefer the transport with the largest BDP (reduces round-trips).
//  3. Penalise transports with ConfidenceNone by 50 % (untrusted estimate).
//
// All methods are safe for concurrent use.
type BandwidthDirector struct {
	mu         sync.RWMutex
	matrix     map[matrixKey]BandwidthSample
	estimators map[string]*BandwidthEstimator // keyed by lower-case transport name
}

type matrixKey struct {
	peer      string
	transport string
}

// defaultPowerCosts mirrors ITransportService conventions (lower = preferred).
var defaultPowerCosts = map[string]int{
	"nearlink":     1,
	"ble":          2,
	"wi-fi direct": 3,
	"circlelink":   3,
	"quic relay":   10,
	"http relay":   10,
}

// NewBandwidthDirector returns a ready-to-use BandwidthDirector.
func NewBandwidthDirector() *BandwidthDirector {
	return &BandwidthDirector{
		matrix:     make(map[matrixKey]BandwidthSample),
		estimators: make(map[string]*BandwidthEstimator),
	}
}

// Register wires an estimator into the director.
// When the estimator fires a SampleImproved callback the director updates every
// known peer's entry for that transport in the matrix.
func (d *BandwidthDirector) Register(estimator *BandwidthEstimator) {
	key := strings.ToLower(estimator.TransportName())

	d.mu.Lock()
	d.estimators[key] = estimator
	d.mu.Unlock()

	estimator.OnSampleImproved(func(s BandwidthSample) {
		d.mu.Lock()
		defer d.mu.Unlock()
		tLower := strings.ToLower(s.TransportName)
		for k := range d.matrix {
			if k.transport == tLower {
				d.matrix[k] = s
			}
		}
	})
}

// GetEstimate returns the bandwidth estimate for a specific peer on a specific transport.
// Returns nil if no estimate exists yet.
func (d *BandwidthDirector) GetEstimate(peerUhid, transportName string) *BandwidthSample {
	key := matrixKey{
		peer:      peerUhid,
		transport: strings.ToLower(transportName),
	}
	d.mu.RLock()
	s, ok := d.matrix[key]
	d.mu.RUnlock()
	if !ok {
		return nil
	}
	return &s
}

// GetEstimates returns all current estimates for a peer across all transports,
// ranked by AvailableBps descending.
func (d *BandwidthDirector) GetEstimates(peerUhid string) []BandwidthSample {
	d.mu.RLock()
	var out []BandwidthSample
	for k, s := range d.matrix {
		if k.peer == peerUhid {
			out = append(out, s)
		}
	}
	d.mu.RUnlock()

	sort.Slice(out, func(i, j int) bool {
		return out[i].AvailableBps > out[j].AvailableBps
	})
	return out
}

// RecommendTransport returns the best transport name for the given peer and payload size.
// Returns an empty string if no transports are registered.
func (d *BandwidthDirector) RecommendTransport(peerUhid string, payloadBytes int64) string {
	candidates := d.GetEstimates(peerUhid)
	if len(candidates) == 0 {
		// No measurement data yet — fall back to lowest power-cost registered transport.
		d.mu.RLock()
		defer d.mu.RUnlock()
		best := ""
		bestCost := int(^uint(0) >> 1) // MaxInt
		for _, e := range d.estimators {
			cost := powerCost(e.TransportName())
			if cost < bestCost {
				bestCost = cost
				best = e.TransportName()
			}
		}
		return best
	}

	var bestName string
	bestScore := -1.0
	for _, s := range candidates {
		pc := float64(powerCost(s.TransportName))
		available := float64(s.AvailableBps)

		// Prefer larger BDP for large payloads.
		bdpBonus := 1.0
		if payloadBytes <= s.BdpBytes {
			bdpBonus = 1.5
		}

		// Penalise untrusted estimates.
		confidenceFactor := 1.0
		if s.Confidence == ConfidenceNone {
			confidenceFactor = 0.5
		}

		score := (available / pc) * bdpBonus * confidenceFactor
		if score > bestScore {
			bestScore = score
			bestName = s.TransportName
		}
	}
	return bestName
}

// BuildGossipPayload constructs a gossip payload to send to a new peer.
// Returns nil if the estimator has ConfidenceNone (no data worth gossiping).
func (d *BandwidthDirector) BuildGossipPayload(peerUhid, transportName string) *BandwidthGossipPayload {
	key := strings.ToLower(transportName)
	d.mu.RLock()
	estimator, ok := d.estimators[key]
	d.mu.RUnlock()
	if !ok {
		return nil
	}

	s := estimator.CurrentSample()
	if s.Confidence == ConfidenceNone {
		return nil
	}

	return &BandwidthGossipPayload{
		PeerUhid:      peerUhid,
		TransportName: transportName,
		BtlBwBps:      s.BtlBwBps,
		RtPropUs:      s.RtProp.Microseconds(),
		Confidence:    s.Confidence,
		MeasuredAt:    s.MeasuredAt,
	}
}

// ApplyGossip receives and applies a gossip payload from a remote peer.
// The appropriate estimator is warmed from the payload, and the matrix is seeded
// so GetEstimate returns something even before probing begins.
func (d *BandwidthDirector) ApplyGossip(payload BandwidthGossipPayload) {
	key := strings.ToLower(payload.TransportName)
	d.mu.RLock()
	estimator, ok := d.estimators[key]
	d.mu.RUnlock()
	if !ok {
		return
	}

	estimator.WarmFromGossip(
		payload.BtlBwBps,
		time.Duration(payload.RtPropUs)*time.Microsecond,
		payload.Confidence,
	)

	// Seed the matrix so GetEstimate returns something immediately.
	mk := matrixKey{peer: payload.PeerUhid, transport: key}
	s := estimator.CurrentSample()
	d.mu.Lock()
	d.matrix[mk] = s
	d.mu.Unlock()
}

// powerCost looks up the relative power cost for a transport name.
// Unknown transports get cost 5 (middle of the range).
func powerCost(transportName string) int {
	if c, ok := defaultPowerCosts[strings.ToLower(transportName)]; ok {
		return c
	}
	return 5
}

File: rust/src/bandwidth/director.rs
// SPDX-License-Identifier: MIT

//! Cross-transport bandwidth synthesis and mesh gossip coordinator.
//!
//! Ports `src/AetherNet.Transport/Bandwidth/BandwidthDirector.cs`.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use super::estimator::BandwidthEstimator;
use super::models::{BandwidthConfidence, BandwidthGossipPayload, BandwidthSample};

// ── Power costs ───────────────────────────────────────────────────────────────

fn default_power_cost(transport: &str) -> f64 {
    match transport {
        "NearLink" => 1.0,
        "BLE" => 2.0,
        "Wi-Fi Direct" | "CircleLink" => 3.0,
        "QUIC Relay" | "HTTP Relay" => 10.0,
        _ => 5.0,
    }
}

// ── Director ──────────────────────────────────────────────────────────────────

/// Cross-transport bandwidth synthesis and mesh gossip coordinator.
///
/// Maintains a (peer_uhid × transport_name) → `BandwidthSample` matrix and
/// provides transport recommendations based on payload size, BDP, and power cost.
pub struct BandwidthDirector {
    // (peer_uhid, transport_name) → latest sample
    matrix: Mutex<HashMap<(String, String), BandwidthSample>>,
    // transport_name → estimator
    estimators: Mutex<HashMap<String, Arc<Mutex<BandwidthEstimator>>>>,
}

impl BandwidthDirector {
    pub fn new() -> Self {
        Self {
            matrix: Mutex::new(HashMap::new()),
            estimators: Mutex::new(HashMap::new()),
        }
    }

    // ── Registration ─────────────────────────────────────────────────────────

    /// Register an estimator. Call once per transport at startup.
    pub fn register(&self, estimator: Arc<Mutex<BandwidthEstimator>>) {
        let transport_name = estimator.lock().unwrap().transport_name.clone();
        // The estimator's live sample is consulted on every query/recommendation (see
        // get_estimate / get_estimates), so a locally-probed transport contributes to
        // selection even before any gossip arrives. This is a pull-based equivalent of
        // the C# reference's estimator-improvement -> matrix propagation (a push closure
        // that borrows `self` is awkward in Rust, so we pull on demand instead).
        self.estimators
            .lock()
            .unwrap()
            .insert(transport_name, estimator);
    }

    // ── Estimates ─────────────────────────────────────────────────────────────

    /// Get the bandwidth estimate for a specific peer on a specific transport.
    /// Returns `None` if no estimate exists yet.
    ///
    /// Reads the gossiped (peer x transport) matrix first; if there is no gossiped
    /// sample for this pair it falls back to the live local estimator for `transport`
    /// (when registered and already probed to a confident sample). This mirrors the C#
    /// reference, where estimator improvements propagate into the matrix.
    pub fn get_estimate(&self, peer_uhid: &str, transport: &str) -> Option<BandwidthSample> {
        {
            let m = self.matrix.lock().unwrap();
            if let Some(s) = m
                .iter()
                .find(|((peer, t), _)| {
                    peer.eq_ignore_ascii_case(peer_uhid) && t.eq_ignore_ascii_case(transport)
                })
                .map(|(_, s)| s.clone())
            {
                return Some(s);
            }
        }
        let g = self.estimators.lock().unwrap();
        g.iter()
            .find(|(name, _)| name.eq_ignore_ascii_case(transport))
            .map(|(_, est)| est.lock().unwrap().current_sample())
            .filter(|s| s.confidence != BandwidthConfidence::None)
    }

    /// Get all current estimates for a peer across all transports,
    /// ranked by `available_bps` descending.
    ///
    /// Merges the gossiped matrix for the peer with the live local-estimator samples for
    /// any probed transport not already covered by a gossiped entry, so a locally-measured
    /// transport is a selection candidate before first gossip (mirrors the C# reference's
    /// estimator -> matrix propagation). An un-probed transport (confidence None) is left
    /// out so a node with no measurements still falls through to the lowest-power default.
    pub fn get_estimates(&self, peer_uhid: &str) -> Vec<BandwidthSample> {
        let mut results: Vec<BandwidthSample> = Vec::new();
        let mut covered: std::collections::HashSet<String> = std::collections::HashSet::new();
        {
            let m = self.matrix.lock().unwrap();
            for ((peer, t), s) in m.iter() {
                if peer.eq_ignore_ascii_case(peer_uhid) {
                    results.push(s.clone());
                    covered.insert(t.to_ascii_lowercase());
                }
            }
        }
        {
            let g = self.estimators.lock().unwrap();
            for (name, est) in g.iter() {
                if !covered.contains(&name.to_ascii_lowercase()) {
                    let s = est.lock().unwrap().current_sample();
                    if s.confidence != BandwidthConfidence::None {
                        results.push(s);
                    }
                }
            }
        }
        results.sort_by(|a, b| b.available_bps.cmp(&a.available_bps));
        results
    }

    /// Recommend the best transport for a payload of `payload_bytes`.
    ///
    /// Scoring (higher = better):
    /// 1. `score = (available_bps / power_cost) × bdp_bonus × confidence_factor`
    /// 2. `bdp_bonus = 1.5` when payload fits in BDP, else 1.0.
    /// 3. `confidence_factor = 0.5` for `None` confidence, else 1.0.
    pub fn recommend_transport(&self, peer_uhid: &str, payload_bytes: i64) -> Option<String> {
        let candidates = self.get_estimates(peer_uhid);
        if candidates.is_empty() {
            // No measurement data yet — fall back to lowest power-cost estimator.
            let g = self.estimators.lock().unwrap();
            return g
                .keys()
                .min_by(|a, b| {
                    default_power_cost(a)
                        .partial_cmp(&default_power_cost(b))
                        .unwrap_or(std::cmp::Ordering::Equal)
                })
                .cloned();
        }

        let mut best: Option<String> = None;
        let mut best_score = f64::NEG_INFINITY;

        for s in &candidates {
            let power_cost = default_power_cost(&s.transport_name);
            let available = s.available_bps as f64;
            let bdp_bonus = if payload_bytes <= s.bdp_bytes { 1.5 } else { 1.0 };
            let confidence_factor = if s.confidence == BandwidthConfidence::None {
                0.5
            } else {
                1.0
            };
            let score = (available / power_cost) * bdp_bonus * confidence_factor;
            if score > best_score {
                best_score = score;
                best = Some(s.transport_name.clone());
            }
        }
        best
    }

    // ── Gossip ────────────────────────────────────────────────────────────────

    /// Build a gossip payload for a new peer that has just completed handshake.
    /// Returns `None` if the estimator has no confident estimate yet.
    pub fn build_gossip_payload(
        &self,
        peer_uhid: &str,
        transport_name: &str,
    ) -> Option<BandwidthGossipPayload> {
        let est = self.estimators.lock().unwrap().get(transport_name).cloned()?;
        let sample = est.lock().unwrap().current_sample();
        if sample.confidence == BandwidthConfidence::None {
            return None;
        }
        Some(BandwidthGossipPayload {
            peer_uhid: peer_uhid.to_string(),
            transport_name: transport_name.to_string(),
            btl_bw_bps: sample.btl_bw_bps,
            rt_prop_us: sample.rt_prop.as_micros() as i64,
            confidence: sample.confidence,
            measured_at: sample.measured_at,
        })
    }

    /// Receive and apply a gossip payload from a remote peer.
    pub fn apply_gossip(&self, payload: BandwidthGossipPayload) {
        let est = {
            let g = self.estimators.lock().unwrap();
            g.get(&payload.transport_name).cloned()
        };
        if let Some(arc) = est {
            let rt_prop = std::time::Duration::from_micros(payload.rt_prop_us.max(0) as u64);
            arc.lock()
                .unwrap()
                .warm_from_gossip(payload.btl_bw_bps, rt_prop, payload.confidence);
            // Seed matrix so get_estimate returns something before first probe.
            let sample = arc.lock().unwrap().current_sample();
            self.matrix
                .lock()
                .unwrap()
                .insert((payload.peer_uhid, payload.transport_name), sample);
        }
    }
}

impl Default for BandwidthDirector {
    fn default() -> Self {
        Self::new()
    }
}
